### PR TITLE
dvc/tree/local: fix it for no repo/cache case

### LINKED ISSUE #5069: get-url: NPE on downloading a directory
# Bug Report

## Description

` dvc get-url  -v s3://dvc-temp/ivan/ docs`

leads to:

```
AttributeError: 'NoneType' object has no attribute 'cache'
```

(see the full stack trace in the Details section below)

### Reproduce

Run the command above. Path on S3 should contain directories.

### Expected

A dir downloaded.

### Environment information

**Output of `dvc version`:**

```console
DVC version: 1.11.0+13bb56
---------------------------------
Platform: Python 3.8.6 on macOS-10.15.6-x86_64-i386-64bit
Supports: All remotes
```

**Additional Information (if any):**

<details>

Full log:

```
2020-12-09 12:24:32,622 DEBUG: Downloading 's3://dvc-temp/ivan/docs/README' to 'docs/docs/README'
2020-12-09 12:24:32,624 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/filelist.doctree' to 'docs/docs/build/doctrees/filelist.doctree'
2020-12-09 12:24:32,624 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/filemanagement.doctree' to 'docs/docs/build/doctrees/filemanagement.doctree'
2020-12-09 12:24:32,624 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/oauth.doctree' to 'docs/docs/build/doctrees/oauth.doctree'
2020-12-09 12:24:32,629 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/quickstart.doctree' to 'docs/docs/build/doctrees/quickstart.doctree'
2020-12-09 12:24:32,629 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/environment.pickle' to 'docs/docs/build/doctrees/environment.pickle'
2020-12-09 12:24:32,630 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/.buildinfo' to 'docs/docs/build/html/.buildinfo'
2020-12-09 12:24:32,630 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/pydrive2.doctree' to 'docs/docs/build/doctrees/pydrive2.doctree'
2020-12-09 12:24:32,630 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/doctrees/index.doctree' to 'docs/docs/build/doctrees/index.doctree'
2020-12-09 12:24:32,635 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_sources/filelist.rst.txt' to 'docs/docs/build/html/_sources/filelist.rst.txt'
2020-12-09 12:24:32,635 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_sources/index.rst.txt' to 'docs/docs/build/html/_sources/index.rst.txt'
2020-12-09 12:24:32,635 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_sources/filemanagement.rst.txt' to 'docs/docs/build/html/_sources/filemanagement.rst.txt'
2020-12-09 12:24:32,635 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_sources/oauth.rst.txt' to 'docs/docs/build/html/_sources/oauth.rst.txt'
2020-12-09 12:24:32,641 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_sources/pydrive2.rst.txt' to 'docs/docs/build/html/_sources/pydrive2.rst.txt'
2020-12-09 12:24:32,641 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_sources/quickstart.rst.txt' to 'docs/docs/build/html/_sources/quickstart.rst.txt'
2020-12-09 12:24:32,641 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/basic.css' to 'docs/docs/build/html/_static/basic.css'
2020-12-09 12:24:32,643 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/badge_only.css' to 'docs/docs/build/html/_static/css/badge_only.css'
2020-12-09 12:24:32,643 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/Roboto-Slab-Bold.woff' to 'docs/docs/build/html/_static/css/fonts/Roboto-Slab-Bold.woff'
2020-12-09 12:24:32,648 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/Roboto-Slab-Bold.woff2' to 'docs/docs/build/html/_static/css/fonts/Roboto-Slab-Bold.woff2'
2020-12-09 12:24:32,648 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/Roboto-Slab-Regular.woff' to 'docs/docs/build/html/_static/css/fonts/Roboto-Slab-Regular.woff'
2020-12-09 12:24:32,649 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/Roboto-Slab-Regular.woff2' to 'docs/docs/build/html/_static/css/fonts/Roboto-Slab-Regular.woff2'
2020-12-09 12:24:32,653 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/fontawesome-webfont.eot' to 'docs/docs/build/html/_static/css/fonts/fontawesome-webfont.eot'
2020-12-09 12:24:32,653 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/fontawesome-webfont.svg' to 'docs/docs/build/html/_static/css/fonts/fontawesome-webfont.svg'
2020-12-09 12:24:32,654 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/fontawesome-webfont.ttf' to 'docs/docs/build/html/_static/css/fonts/fontawesome-webfont.ttf'
2020-12-09 12:24:32,658 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/fontawesome-webfont.woff2' to 'docs/docs/build/html/_static/css/fonts/fontawesome-webfont.woff2'
2020-12-09 12:24:32,658 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/fontawesome-webfont.woff' to 'docs/docs/build/html/_static/css/fonts/fontawesome-webfont.woff'
2020-12-09 12:24:32,659 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-bold-italic.woff2' to 'docs/docs/build/html/_static/css/fonts/lato-bold-italic.woff2'
2020-12-09 12:24:32,659 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-bold-italic.woff' to 'docs/docs/build/html/_static/css/fonts/lato-bold-italic.woff'
2020-12-09 12:24:32,663 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-bold.woff' to 'docs/docs/build/html/_static/css/fonts/lato-bold.woff'
2020-12-09 12:24:32,664 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-bold.woff2' to 'docs/docs/build/html/_static/css/fonts/lato-bold.woff2'
2020-12-09 12:24:32,672 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-normal-italic.woff' to 'docs/docs/build/html/_static/css/fonts/lato-normal-italic.woff'
2020-12-09 12:24:32,672 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-normal-italic.woff2' to 'docs/docs/build/html/_static/css/fonts/lato-normal-italic.woff2'
2020-12-09 12:24:33,019 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-normal.woff' to 'docs/docs/build/html/_static/css/fonts/lato-normal.woff'
2020-12-09 12:24:33,034 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/fonts/lato-normal.woff2' to 'docs/docs/build/html/_static/css/fonts/lato-normal.woff2'
2020-12-09 12:24:33,061 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/css/theme.css' to 'docs/docs/build/html/_static/css/theme.css'
2020-12-09 12:24:33,066 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/doctools.js' to 'docs/docs/build/html/_static/doctools.js'
2020-12-09 12:24:33,076 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/documentation_options.js' to 'docs/docs/build/html/_static/documentation_options.js'
2020-12-09 12:24:33,077 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/FontAwesome.otf' to 'docs/docs/build/html/_static/fonts/FontAwesome.otf'
2020-12-09 12:24:33,077 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/file.png' to 'docs/docs/build/html/_static/file.png'
2020-12-09 12:24:33,086 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bold.woff2' to 'docs/docs/build/html/_static/fonts/Lato/lato-bold.woff2'
2020-12-09 12:24:33,087 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bolditalic.eot' to 'docs/docs/build/html/_static/fonts/Lato/lato-bolditalic.eot'
2020-12-09 12:24:33,090 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bold.ttf' to 'docs/docs/build/html/_static/fonts/Lato/lato-bold.ttf'
2020-12-09 12:24:33,090 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bold.eot' to 'docs/docs/build/html/_static/fonts/Lato/lato-bold.eot'
2020-12-09 12:24:33,090 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bold.woff' to 'docs/docs/build/html/_static/fonts/Lato/lato-bold.woff'
2020-12-09 12:24:33,096 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bolditalic.ttf' to 'docs/docs/build/html/_static/fonts/Lato/lato-bolditalic.ttf'
2020-12-09 12:24:33,112 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bolditalic.woff' to 'docs/docs/build/html/_static/fonts/Lato/lato-bolditalic.woff'
2020-12-09 12:24:33,112 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-bolditalic.woff2' to 'docs/docs/build/html/_static/fonts/Lato/lato-bolditalic.woff2'
2020-12-09 12:24:33,128 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-italic.eot' to 'docs/docs/build/html/_static/fonts/Lato/lato-italic.eot'
2020-12-09 12:24:33,155 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-italic.ttf' to 'docs/docs/build/html/_static/fonts/Lato/lato-italic.ttf'
2020-12-09 12:24:33,171 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-italic.woff' to 'docs/docs/build/html/_static/fonts/Lato/lato-italic.woff'
2020-12-09 12:24:33,180 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-regular.eot' to 'docs/docs/build/html/_static/fonts/Lato/lato-regular.eot'
2020-12-09 12:24:33,180 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-regular.ttf' to 'docs/docs/build/html/_static/fonts/Lato/lato-regular.ttf'
2020-12-09 12:24:33,180 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-italic.woff2' to 'docs/docs/build/html/_static/fonts/Lato/lato-italic.woff2'
2020-12-09 12:24:33,187 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-regular.woff' to 'docs/docs/build/html/_static/fonts/Lato/lato-regular.woff'
2020-12-09 12:24:33,198 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Lato/lato-regular.woff2' to 'docs/docs/build/html/_static/fonts/Lato/lato-regular.woff2'
2020-12-09 12:24:33,224 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Bold.woff' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Bold.woff'
2020-12-09 12:24:33,231 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Bold.woff2' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Bold.woff2'
2020-12-09 12:24:33,234 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Light.woff' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Light.woff'
2020-12-09 12:24:33,247 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Light.woff2' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Light.woff2'
2020-12-09 12:24:33,270 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Regular.woff' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Regular.woff'
2020-12-09 12:24:33,296 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Regular.woff2' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Regular.woff2'
2020-12-09 12:24:33,313 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Thin.woff' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Thin.woff'
2020-12-09 12:24:33,319 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/Roboto-Slab-Thin.woff2' to 'docs/docs/build/html/_static/fonts/Roboto-Slab-Thin.woff2'
2020-12-09 12:24:33,321 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.eot' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.eot'
2020-12-09 12:24:33,341 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.ttf' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.ttf'
2020-12-09 12:24:33,342 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.woff' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.woff'
2020-12-09 12:24:33,350 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.woff2' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-bold.woff2'
2020-12-09 12:24:33,366 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.ttf' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.ttf'
2020-12-09 12:24:33,366 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.eot' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.eot'
2020-12-09 12:24:33,389 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.woff' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.woff'
2020-12-09 12:24:33,425 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.woff2' to 'docs/docs/build/html/_static/fonts/RobotoSlab/roboto-slab-v7-regular.woff2'
2020-12-09 12:24:33,433 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/fontawesome-webfont.eot' to 'docs/docs/build/html/_static/fonts/fontawesome-webfont.eot'
2020-12-09 12:24:33,460 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/fontawesome-webfont.svg' to 'docs/docs/build/html/_static/fonts/fontawesome-webfont.svg'
2020-12-09 12:24:33,464 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/fontawesome-webfont.ttf' to 'docs/docs/build/html/_static/fonts/fontawesome-webfont.ttf'
2020-12-09 12:24:33,471 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/fontawesome-webfont.woff' to 'docs/docs/build/html/_static/fonts/fontawesome-webfont.woff'
2020-12-09 12:24:33,553 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/fontawesome-webfont.woff2' to 'docs/docs/build/html/_static/fonts/fontawesome-webfont.woff2'
2020-12-09 12:24:33,556 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-bold-italic.woff' to 'docs/docs/build/html/_static/fonts/lato-bold-italic.woff'
2020-12-09 12:24:33,557 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-bold-italic.woff2' to 'docs/docs/build/html/_static/fonts/lato-bold-italic.woff2'
2020-12-09 12:24:33,561 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-bold.woff' to 'docs/docs/build/html/_static/fonts/lato-bold.woff'
2020-12-09 12:24:33,574 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-bold.woff2' to 'docs/docs/build/html/_static/fonts/lato-bold.woff2'
2020-12-09 12:24:33,575 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-normal-italic.woff2' to 'docs/docs/build/html/_static/fonts/lato-normal-italic.woff2'
2020-12-09 12:24:33,575 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-normal-italic.woff' to 'docs/docs/build/html/_static/fonts/lato-normal-italic.woff'
2020-12-09 12:24:33,585 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-normal.woff' to 'docs/docs/build/html/_static/fonts/lato-normal.woff'
2020-12-09 12:24:33,585 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/fonts/lato-normal.woff2' to 'docs/docs/build/html/_static/fonts/lato-normal.woff2'
2020-12-09 12:24:33,587 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/jquery-3.5.1.js' to 'docs/docs/build/html/_static/jquery-3.5.1.js'
2020-12-09 12:24:33,588 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/jquery.js' to 'docs/docs/build/html/_static/jquery.js'
2020-12-09 12:24:33,590 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/js/badge_only.js' to 'docs/docs/build/html/_static/js/badge_only.js'
2020-12-09 12:24:33,590 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/js/html5shiv.min.js' to 'docs/docs/build/html/_static/js/html5shiv.min.js'
2020-12-09 12:24:33,590 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/js/html5shiv-printshiv.min.js' to 'docs/docs/build/html/_static/js/html5shiv-printshiv.min.js'
2020-12-09 12:24:33,591 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/js/modernizr.min.js' to 'docs/docs/build/html/_static/js/modernizr.min.js'
2020-12-09 12:24:33,729 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/js/theme.js' to 'docs/docs/build/html/_static/js/theme.js'
2020-12-09 12:24:33,748 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/minus.png' to 'docs/docs/build/html/_static/minus.png'
2020-12-09 12:24:33,748 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/language_data.js' to 'docs/docs/build/html/_static/language_data.js'
2020-12-09 12:24:33,750 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/pygments.css' to 'docs/docs/build/html/_static/pygments.css'
2020-12-09 12:24:33,754 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/plus.png' to 'docs/docs/build/html/_static/plus.png'
2020-12-09 12:24:33,756 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/searchtools.js' to 'docs/docs/build/html/_static/searchtools.js'
2020-12-09 12:24:33,758 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/underscore-1.3.1.js' to 'docs/docs/build/html/_static/underscore-1.3.1.js'
2020-12-09 12:24:33,758 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/_static/underscore.js' to 'docs/docs/build/html/_static/underscore.js'
2020-12-09 12:24:33,770 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/filelist.html' to 'docs/docs/build/html/filelist.html'
2020-12-09 12:24:33,784 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/genindex.html' to 'docs/docs/build/html/genindex.html'
2020-12-09 12:24:33,784 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/filemanagement.html' to 'docs/docs/build/html/filemanagement.html'
2020-12-09 12:24:33,790 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/index.html' to 'docs/docs/build/html/index.html'
2020-12-09 12:24:33,807 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/oauth.html' to 'docs/docs/build/html/oauth.html'
2020-12-09 12:24:33,811 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/objects.inv' to 'docs/docs/build/html/objects.inv'
2020-12-09 12:24:33,833 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/py-modindex.html' to 'docs/docs/build/html/py-modindex.html'
2020-12-09 12:24:33,841 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/pydrive2.html' to 'docs/docs/build/html/pydrive2.html'
2020-12-09 12:24:33,920 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/quickstart.html' to 'docs/docs/build/html/quickstart.html'
2020-12-09 12:24:33,921 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/searchindex.js' to 'docs/docs/build/html/searchindex.js'
2020-12-09 12:24:33,921 DEBUG: Downloading 's3://dvc-temp/ivan/docs/build/html/search.html' to 'docs/docs/build/html/search.html'
2020-12-09 12:24:33,936 DEBUG: Downloading 's3://dvc-temp/ivan/docs/conf.py' to 'docs/docs/conf.py'
2020-12-09 12:24:33,959 DEBUG: Downloading 's3://dvc-temp/ivan/docs/filemanagement.rst' to 'docs/docs/filemanagement.rst'
2020-12-09 12:24:33,959 DEBUG: Downloading 's3://dvc-temp/ivan/docs/filelist.rst' to 'docs/docs/filelist.rst'
2020-12-09 12:24:33,990 DEBUG: Downloading 's3://dvc-temp/ivan/docs/index.rst' to 'docs/docs/index.rst'
2020-12-09 12:24:33,999 DEBUG: Downloading 's3://dvc-temp/ivan/docs/pydrive2.rst' to 'docs/docs/pydrive2.rst'
2020-12-09 12:24:33,999 DEBUG: Downloading 's3://dvc-temp/ivan/docs/oauth.rst' to 'docs/docs/oauth.rst'
2020-12-09 12:24:34,006 DEBUG: Downloading 's3://dvc-temp/ivan/docs/quickstart.rst' to 'docs/docs/quickstart.rst'
2020-12-09 12:24:34,324 ERROR: unexpected error - 'NoneType' object has no attribute 'cache'
------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ivan/Projects/dvc/dvc/main.py", line 90, in main
    ret = cmd.run()
  File "/Users/ivan/Projects/dvc/dvc/command/get_url.py", line 17, in run
    Repo.get_url(self.args.url, out=self.args.out)
  File "/Users/ivan/Projects/dvc/dvc/repo/get_url.py", line 21, in get_url
    out.save()
  File "/Users/ivan/Projects/dvc/dvc/output/base.py", line 276, in save
    self.hash_info = self.get_hash()
  File "/Users/ivan/Projects/dvc/dvc/output/base.py", line 186, in get_hash
    return self.tree.get_hash(self.path_info)
  File "/Users/ivan/Projects/dvc/.env/lib/python3.8/site-packages/funcy/decorators.py", line 39, in wrapper
    return deco(call, *dargs, **dkwargs)
  File "/Users/ivan/Projects/dvc/dvc/tree/base.py", line 45, in use_state
    return call()
  File "/Users/ivan/Projects/dvc/.env/lib/python3.8/site-packages/funcy/decorators.py", line 60, in __call__
    return self._func(*self._args, **self._kwargs)
  File "/Users/ivan/Projects/dvc/dvc/tree/base.py", line 292, in get_hash
    hash_info = self.get_dir_hash(path_info, **kwargs)
  File "/Users/ivan/Projects/dvc/.env/lib/python3.8/site-packages/funcy/decorators.py", line 39, in wrapper
    return deco(call, *dargs, **dkwargs)
  File "/Users/ivan/Projects/dvc/dvc/tree/base.py", line 45, in use_state
    return call()
  File "/Users/ivan/Projects/dvc/.env/lib/python3.8/site-packages/funcy/decorators.py", line 60, in __call__
    return self._func(*self._args, **self._kwargs)
  File "/Users/ivan/Projects/dvc/dvc/tree/base.py", line 352, in get_dir_hash
    hash_info = self.repo.cache.local.save_dir_info(dir_info)
AttributeError: 'NoneType' object has no attribute 'cache'
------------------------------------------------------------
2020-12-09 12:24:40,727 DEBUG: Version info for developers:
DVC version: 1.11.0+13bb56
---------------------------------
Platform: Python 3.8.6 on macOS-10.15.6-x86_64-i386-64bit
Supports: All remotes
```

</details>

### PR DESCRIPTION
**Pull request by dkropachev**
fixes #5069

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] ~~📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.~~

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: tests/func/test_get_url.py [hunk 1]
```diff
@@ -13,6 +13,16 @@ def test_get_file(tmp_dir):
     assert (tmp_dir / "foo_imported").read_text() == "foo contents"
 
 
+def test_get_dir(tmp_dir):
```
Comment (pared) [inline tests/func/test_get_url.py:16]: Should we mark that the purpose of this test is to check no-cache use case?

Comment (dkropachev) [reply tests/func/test_get_url.py:16]: Pwerl,
Before we go to the tests, could you please confirm that fix is done in correct place and in corret manner.

Answering your question:
> Should we mark that the purpose of this test is to check no-cache use case?

This test-suit targets high-level function, get_url, on this level there is no concept of cache, it goes only on lower levels, so we should not mention cache here at all.

Also it is probably needed to have another test for no repo/cache case in test_repo, test_run_cache, test_dvc suits where we could go into "nocache" details.
Let me know what you think.

File: dvc/tree/local.py [hunk 2]
```diff
@@ -316,6 +316,15 @@ def get_file_hash(self, path_info):
...
         return hash_info
 
+    def get_dir_hash_no_cache(self, dir_info):
+        hash_info = HashInfo(
+            self.PARAM_CHECKSUM, dict_md5(dir_info.to_list()),
+        )
+        hash_info.value += self.CHECKSUM_DIR_SUFFIX
+        hash_info.dir_info = dir_info
+        hash_info.nfiles = dir_info.nfiles
+        return hash_info
```
Comment (pared) [inline dvc/tree/local.py:326]: Ok, so it seems to me that we are duplicating code here. There is already code responsible for `dir` hash computations, in `cache/base.py`. Duplicating it here, exposes us to possibility, that we will calculate dir hash differently than we do it when `repo` is present. I think that should not happen, we should reuse the code. I do get that in that case we don't have `repo` to call. Worth noting that we have commands (`plots`, `metrics`) that don't require repo to execute.

**Review by pared — CHANGES_REQUESTED** [review]

File: dvc/tree/local.py (hunk 2 above)
Comment (dkropachev) [reply dvc/tree/local.py]: Redone, please take a look

**Review by pared — COMMENTED** [review]
Looks much better @dkropachev!

Still I think we might make it even better: The error is caused by `out.save` call in `get_url` methdod. I am not sure whether we need it anymore. Could you check it?

**Comment by dkropachev** [comment]
> Looks much better @dkropachev!
> 
> Still I think we might make it even better: The error is caused by `out.save` call in `get_url` methdod. I am not sure whether we need it anymore. Could you check it?

Confirm that out is going to be created with no cache and repo in any case and out.save would endup in doing nothing anyways.
Removed out.save from get_url, do you still want to keep fix to the save() ?

File: dvc/output/base.py [hunk 3]
```diff
@@ -277,7 +277,8 @@ def save(self):
             self.verify_metric()
 
         if not self.use_cache:
-            self.hash_info = self.get_hash()
+            if self.repo:
+                self.hash_info = self.get_hash()
```
Comment (efiop) [inline dvc/output/base.py:281]: From https://github.com/iterative/dvc/pull/5118#issuecomment-750709994 . Let's not intoroduce this change, as it is not necessary.
```suggestion
```

File: dvc/output/base.py [hunk 4]
```diff
@@ -277,7 +277,6 @@ def save(self):
             self.verify_metric()
 
         if not self.use_cache:
-            self.hash_info = self.get_hash()
             if not self.IS_DEPENDENCY:
```
Comment (efiop) [inline dvc/output/base.py:280]:
```suggestion
            self.hash_info = self.get_hash()
            if not self.IS_DEPENDENCY:
```

**Review by efiop — APPROVED** [review]
@dkropachev Thank you for the PR! :pray:

And thanks to @pared for the review and guidance! :pray:

**Pull request merged**